### PR TITLE
fix: destructuring in catch clause in `no-unused-vars`

### PR DESCRIPTION
**Pull request by fasttime**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Tell us about your environment (`npx eslint --env-info`):**

Node version: v22.3.0
npm version: v10.8.1
Local ESLint version: v9.6.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.5.0

**What parser are you using (place an "X" next to just one item)?**

[X] `Default (Espree)`
[ ] `@typescript-eslint/parser`
[ ] `@babel/eslint-parser`
[ ] `vue-eslint-parser`
[ ] `@angular-eslint/template-parser`
[ ] `Other`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
// All config is inline
export default [];
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
/*eslint no-unused-vars: ["error", {
    "caughtErrorsIgnorePattern": "message",
    "destructuredArrayIgnorePattern": "^_",
    "ignoreRestSiblings": true
}]*/

try {
} catch ({ message }) {
}

try {
} catch ({ errors: [_firstError] }) {
}

try {
} catch ({ stack, ...error }) {
  console.log(error);
}
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHtcbiAgICBcImNhdWdodEVycm9yc0lnbm9yZVBhdHRlcm5cIjogXCJtZXNzYWdlXCIsXG4gICAgXCJkZXN0cnVjdHVyZWRBcnJheUlnbm9yZVBhdHRlcm5cIjogXCJeX1wiLFxuICAgIFwiaWdub3JlUmVzdFNpYmxpbmdzXCI6IHRydWVcbn1dKi9cblxudHJ5IHtcbn0gY2F0Y2ggKHsgbWVzc2FnZSB9KSB7XG59XG5cbnRyeSB7XG59IGNhdGNoICh7IGVycm9yczogW19maXJzdEVycm9yXSB9KSB7XG59XG5cbnRyeSB7XG59IGNhdGNoICh7IHN0YWNrLCAuLi5lcnJvciB9KSB7XG4gIGNvbnNvbGUubG9nKGVycm9yKTtcbn0iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

**What did you expect to happen?**

No problems should be reported.

* In the first catch clause, `message` should be allowed because of `"caughtErrorsIgnorePattern": "message"`.
* In the second catch clause, `_firstError` should be allowed because of `"destructuredArrayIgnorePattern": "^_"`.
* In the last catch clause, `stack` should be allowed because of `"ignoreRestSiblings": true`.

**What actually happened? Please include the actual, raw output from ESLint.**

Problems were reported:

```console
8:12  'message' is defined but never used. Allowed unused caught errors must match /message/u.  (no-unused-vars)
12:21 '_firstError' is defined but never used. Allowed unused caught errors must match /message/u.  (no-unused-vars)
16:12 'stack' is defined but never used. Allowed unused caught errors must match /message/u.  (no-unused-vars)
```

#### What changes did you make? (Give an overview)

<s>Fixed the logic that locates identifier names to account for destructured variables in a catch clause.</s> Updated `eslint-scope` to upcoming version 8.0.2; added unit tests.

#### Is there anything you'd like reviewers to focus on?

This bug already existed before #18608 and #18609, it's not a regression.

### CODE REVIEW

**Review by mdjermanovic — COMMENTED** [review]
> ```js
> try {
> } catch ({ message }) {
> }
> 
> try {
> } catch ({ errors: [_firstError] }) {
> }
> 
> try {
> } catch ({ stack, ...error }) {
>   console.log(error);
> }
> ```

The problem in these cases is that `def.name` is `CatchClause#param` node (object pattern) instead of the identifier node? If that's the cause, I believe this is a bug in eslint-scope rather than in this rule.

**Comment by fasttime** [comment]
> > ```js
> > try {
> > } catch ({ message }) {
> > }
> > 
> > try {
> > } catch ({ errors: [_firstError] }) {
> > }
> > 
> > try {
> > } catch ({ stack, ...error }) {
> >   console.log(error);
> > }
> > ```
> 
> The problem in these cases is that `def.name` is `CatchClause#param` node (object pattern) instead of the identifier node? If that's the cause, I believe this is a bug in eslint-scope rather than in this rule.

Yeah, I guess so. Do you think we can just fix this in `eslint-scope` during a minor release, or would that be a breaking change?

**Comment by mdjermanovic** [comment]
I think we can fix this in `eslint-scope` in a minor (or patch?) release. 

@nzakas what do you think?

Per the documentation, `Definition#name` should always be the `Identifier` node:

https://eslint.org/docs/latest/extend/scope-manager-interface#name-1

**Comment by mdjermanovic** [comment]
Here's a PR to fix this in `eslint-scope`: https://github.com/eslint/eslint-scope/pull/127

**Comment by fasttime** [comment]
> Here's a PR to fix this in `eslint-scope`: [eslint/eslint-scope#127](https://github.com/eslint/eslint-scope/pull/127)

Thanks, that fixes it! I think we can merge that PR and just keep the tests I wrote here after `eslint-scope` is released.

File: package.json
Comment (fasttime) [inline package.json:82]:
```suggestion
    "eslint-scope": "^8.0.2",
```

**Comment by fasttime** [comment]
**TODOs before merging**:

- [ ] Commit suggestion https://github.com/eslint/eslint/pull/18636#discussion_r1674375726
- [ ] Delete branch https://github.com/eslint/eslint-scope/tree/prepare-for-eslint-18636

**Review by mdjermanovic — DISMISSED** [review]
LGTM when TODOs are done. Just converting it to draft so that we don't accidentally merge it in the meantime.

**Pull request merged**
